keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2397 @@
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources is too limited. This KEP addresses
+limitations of the current approach for the following use cases:
+
+- *Device initialization*: When starting a workload that uses
+  an accelerator like an FPGA, I’d like to have the accelerator
+  reconfigured or reprogrammed for the workload before the workload
+  itself starts. For security reasons, workloads should not be able to
+  reconfigure devices directly.
+
+  *Limitation*: Currently, it’s impossible to specify the desired
+  device properties that are required for reconfiguring devices.
+  For the FPGA example, a file containing the desired configuration
+  of the FPGA has to be referenced.
+
+- *Device cleanup*: When my workload is finished, I would like to have
+  a mechanism for cleanup of the device, that will ensure that device
+  does not contain traces/parameters/data from previous workloads and
+  appropriate power state/shutdown. For example, an FPGA might have
+  to be reset because its configuration for the workload was
+  confidential.
+
+  *Limitation*: Post-stop actions are not supported.
+
+- *Partial allocation*: When deploying a container I’d like to be able
+  to use part of the shareable device inside a container and other
+  containers should be able to use other free resources on the same
+  device.
+
+  *Limitation*: Current implementation of the device plugin doesn’t
+  allow one to allocate part of the device because parameters are too limited
+  and Kubernetes doesn't have enough information about the extended
+  resources on a node to decide whether they can be shared.
+
+- *Optional allocation*: When deploying a workload I’d like to specify
+  soft(optional) device requirements. If a device exists and it’s
+  allocatable it will be allocated. If not - the workload will be run on
+  a node without a device. GPU and crypto-offload engines are
+  examples of this kind of device. If they’re not available, workloads
+  can still run by falling back to using only the CPU for the same
+  task.
+
+  *Limitation*: Optional allocation is supported neither by the device
+  plugins nor by current Pod resource declaration.
+
+- *Support Over the Fabric devices*: When deploying a container, I’d
+  like to utilize devices available over the Fabric (network, special
+  links, etc).
+
+  *Limitation*: The device plugin API is designed for node-local resources that
+  get discovered by a plugin running on the node. Projects like
+  [Akri](https://www.cncf.io/projects/akri/) have to work around that by
+  reporting the same network-attached resource on all nodes that it could
+  get attached to and then updating resource availability on all of those
+  nodes when resources get used.
+
+Several other limitations are addressed by
+[CDI](https://github.com/container-orchestrated-devices/container-device-interface/),
```

Comment:
What is the status of CDI in the container ecosystem in general? Is it well supported by device vendors (is there some competing standards)? I noticed it is not part of the "official" OCI spec, is that planned?

Reply:
I've never heard of it until now and I already found something that isn't very friendly for kube (env vars)

Reply:
There's a write-up from Nvidia about issues with environment variables and k8s for device management [in the Overview of this document](https://docs.google.com/document/d/1uXVF-NWZQXgP1MLb87_kMkQvidpnkNWicdpO2l9g-fw/edit#heading=h.b3ti65rojfy5) (linked from [NVIDIA/k8s-device-plugin](https://github.com/NVIDIA/k8s-device-plugin)); they note that the intended longer-term fix for that is CDI.

I don't think that env-vars are intrinsic to CDI operation, it's more that CDI _can_ be used to define env-vars to pass into the container, but can also use other mechanisms, and for the situations being discussed in the document I linked, interfacing with the Nvidia k8s-device-plugin would presumably be via OCI spec editing or other mechanism that cannot be done "manually" by a pod-spec to bypass protection mechanisms.

That said, I haven't spent a lot of time on CDI, so perhaps @thockin has realised some problem I have not noticed.

Reply:
I think the key thing that solves the problems noted in the linked document is not CDI as such, but rather the extension of the CRI with explicit fields for specifying devices. This means that this can be used instead of environment variables and annotations (which are also user-editable) to indicate which devices are required.

You are correct that the env vars (and other modifications) that are defined in the CDI spec are never exposed to k8s since the (CDI-enabled) runtime will make the modifications to the OCI spec before passing this on to the low-level runtime such as `runc`.

Reply:
There is an open issue in CDI about this: https://github.com/container-orchestrated-devices/container-device-interface/issues/41

The solution in the context of this KEP could be to have a driver-specific ResourceClaimParameter where the user specifies a prefix (or suffix, both would work) for the env variables that the driver is going to write into the CDI file. Then ResourceClaim "my-big-gpu" could have variables `MY_BIG_GPU_ID` and ResourceClaim "my-fast-gpu" `MY_FAST_GPU_ID`.

This already works, the KEP doesn't need to be changed for it.

Reply:
That makes sense. I meant to come back and clarify that the reason I thought this would be better-discussed on the CDI tracker is because we're either talking about env-vars from the CDI JSON file which are vendor specific (and hence can't be usefully changed/prefixed, etc. e.g. `NVIDIA_VISIBLE_DEVICES` as discussed in that linked ticket, because the existing device hook already knows about it by that exact name), or it's an env-var which must be supported by the Resource Driver, using some kind of downward API-like system as @thockin mooted above, and which could be KEP'd separately if a practical use-case is encountered.

The former type has nothing to do with k8s, it's internal to CDI, Device Plugins support this using `ContainerAllocateResponse.Envs` so doing it with CDI abstracts that away from k8s completely.

The latter type of env-var has nothing to do with CDI nor is it motived by the user stories and goals of this KEP; it _might_ make sense for a general DevicePlugin Downward API to allow populating ContainerSpec env-vars or volumes from arbitrary CDI-sourced details, or to generically describe how to add custom-named env-vars to the CDI JSON file before it's applied (which would require that a CDI file is _generated_ in `NodePrepareResource`, currently that's not the case, it could be statically created for e.g., the Windows DirectX class GUID device) but I would nail this design down first before trying to add a feature like that since it seems purely additive.

Reply:
As a record, not the blocker for moving forward, we debated on this topic (implicit vs explicit API to K8s) while developing today's device plugin API, and here is what I roughly remembered: 

Back then (~2017?), device vendor, especially Nvidia would like an implicit API model to encode special environment variables in the container images to GPU users, and have the special container runtime hook to interpret those environment variables and change container runtime settings accordingly. The reasons from the device vendor are straightforward using GPU as an example: 

- The same software can be reused in various container running environments, whether it is through single-machine container runtime command or it is through cluster-level container orchestration platform like Kubernetes.
- It allows faster iteration to develop and release new features from device vendor side, as they have complete control on the container images they provide to GPU customers and the low-level runtime hook they develop and release.

All above are making sense, hence there is CDI effort was initiated. One of the reasons on CDI discussion was pushed down to OCI layer, not CRI is to support single-machine without K8s use case. CRI wasn't merged into Containerd yet back then. We can re-open this topic now if needed. 

Meanwhile SIG Node, representing K8s community, raised the concerns on the visibility, maintainability, policy control, backward/forward compatibility for K8s vendors. Similar issues remain in CSI too.

Reply:
As @dchen1107 mentions, once CDI is adopted as the standard way to inject "abstract" devices into a container (i.e. by using a single identifier to trigger the injection of multiple device nodes, binaries, libs, etc.), then there will be no need for using an environment variable to accomplish the same thing. CDI then replaces the awkward, unstructured envvar approach used by NVIDIA today in favour of a well-defined, CRI integrated way of passing the required information down the stack.

Reply:
I'm happy to take the discussion somewhere else - my main point is this:

Almost everything that we put into a running container is defined by the user.  They bring their own image.  They specifiy where to mount volumes.  They specify environment variables. We almost never do those things unilaterally and when we do, we have mostly regretted it.

> If the device-usage support inside the container depends on a vendor-defined env-var for discovery, why would we want to force the Pod-creator to explicitly populate that env-var as well as selecting the device via a Device Plugin or this KEP?

With the caveat that I am not steeped in this space - I argue that there should not be such a thing as a vendor-defined env-var in the context of a user's container.  The user defines the env vars - we just offer them the values.  If you have things that look for hardcoded env var names, you're (usually) doing your users a disservice.  What they see inside the running container should be up to them to define.

We've gone to great lengths to NOT do that.